PublicComponentCode/ComponentCode/src/java/main/ccsu/edu/grovepicomponents/TemperatureAndHumiditySensor.java

```diff
@@ -0,0 +1,132 @@
+package ccsu.edu.grovepicomponents;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import edu.ccsu.error.IncompatibleSensorError;
+import edu.ccsu.interfaces.Iterator;
+import edu.ccsu.utility.CommonConstants;
+
+/***
+ * Temperature and humidity sensor that reads data from
+ * the GrovePi temp/humidity sensor and returns that data
+ *
+ */
+public class TemperatureAndHumiditySensor extends Sensor {
+
+	private List<TempAndHumidityData> sensorData;
+	/**
+	 * Creates a new Temperature and Humidity Sensor
+	 * @param name
+	 * @param portNumber
+	 */
+	public TemperatureAndHumiditySensor(String name, String portNumber) {
+		this.name = name;
+		this.portNumber = portNumber;
+		this.sensorData = new ArrayList<>();
+		this.sensorFile = CommonConstants.TEMPHUMIDITY;
+	}
+    
+	/**
+	 * Method that takes data string, parses it, creates LightSensorData objects,
+	 * and adds them to list
+	 * @param data
+	 */
+	@Override
+	protected String addToList(String data) {
+		StringBuilder builder = new StringBuilder();
+		String[] dataToAdd = data.split(",");
+		for(String str: dataToAdd) {
+			String[] makeIntoData = str.split(" ");
+			String fahrenheit = makeIntoData[0];
+			String celcius = makeIntoData[1];
+			String humidityPercentage = makeIntoData[2];
+			Date date = new Date();
+			builder.append("Fahrenheit: " + fahrenheit + " Celcius: " + celcius + " HumidityPercentage: " + humidityPercentage + " Date: "+ date + "\n");
+			//value from output will be three numbers, first is fahrenheit, second is celcius, third is humidity percentage
+			sensorData.add(new TempAndHumidityData(Float.parseFloat(fahrenheit), Float.parseFloat(celcius), Float.parseFloat(humidityPercentage), date));
+		}
+		return builder.toString();
+	}
+	
+	@Override
+	protected boolean checkPort(String portNumber) {
+		if(!this.getPortNumber().contains("D")) {
+			System.out.println("Must use a digital port starting with D");
+			return false;
+		}
+		return true;
+	}
+	
+	@Override
+	public void setNextSensor(Sensor nextSensor) throws IncompatibleSensorError {
+		if(nextSensor instanceof TemperatureAndHumiditySensor) {
+			this.nextSensor = nextSensor;
+		} else {
+			throw new IncompatibleSensorError("Sensor not compatible with sensor. Sensor chain can only be use other Sensors");
+		}
+	}
+    
+	@Override
+	public Iterator getIterator() {
```

Comment:
Good use of Iterator protecting visibility, but have the Iterator returned type bound.  Your getIterator method doesn't need to be public.  If it is only used internally so you want to bind it to the protected data type then just change your visibility here to either private or protected and bind it to the internal class.